pkg/generators/merge.go

```diff
@@ -0,0 +1,192 @@
+
+type MergeGenerator struct {
+	// The inner generators supported by the merge generator (cluster, git, list...)
+	supportedGenerators map[string]Generator
+}
+
+func NewMergeGenerator(supportedGenerators map[string]Generator) Generator {
+	m := &MergeGenerator{
+		supportedGenerators: supportedGenerators,
+	}
+	return m
+}
+
+func (m *MergeGenerator) getParamSetsForAllGenerators(generators []argoprojiov1alpha1.ApplicationSetNestedGenerator, appSet *argoprojiov1alpha1.ApplicationSet) ([][]map[string]string, error) {
+	var paramSets [][]map[string]string
+	for _, generator := range generators {
+		generatorParamSets, err := m.getParams(generator, appSet)
+		if err != nil {
+			return nil, err
+		}
+		// concatenate param lists produced by each generator
+		paramSets = append(paramSets, generatorParamSets)
+	}
+	return paramSets, nil
+}
+
+func (m *MergeGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.ApplicationSetGenerator, appSet *argoprojiov1alpha1.ApplicationSet) ([]map[string]string, error) {
+	if appSetGenerator.Merge == nil {
+		return nil, nil
+	}
+
+	if len(appSetGenerator.Merge.Generators) < 2 {
+		return nil, LessThanTwoGeneratorsInMerge
+	}
+
+	paramSetsFromGenerators, err := m.getParamSetsForAllGenerators(appSetGenerator.Merge.Generators, appSet)
+	if err != nil {
+		return nil, err
+	}
+
+	baseParamSetsByMergeKey, err := getParamSetsByMergeKey(appSetGenerator.Merge.MergeKeys, paramSetsFromGenerators[0])
+	if err != nil {
+		return nil, err
+	}
+
+	for _, paramSets := range paramSetsFromGenerators[1:] {
+		paramSetsByMergeKey, err := getParamSetsByMergeKey(appSetGenerator.Merge.MergeKeys, paramSets)
+		if err != nil {
+			return nil, err
+		}
+
+		for mergeKeyValue, baseParamSet := range baseParamSetsByMergeKey {
+			if overrideParamSet, exists := paramSetsByMergeKey[mergeKeyValue]; exists {
+				overriddenParamSet, err := utils.CombineStringMapsAllowDuplicates(baseParamSet, overrideParamSet)
+				if err != nil {
+					return nil, err
+				}
+				baseParamSetsByMergeKey[mergeKeyValue] = overriddenParamSet
+			}
+		}
+	}
+
+	mergedParamSets := make([]map[string]string, len(baseParamSetsByMergeKey))
+	var i = 0
+	for _, mergedParamSet := range baseParamSetsByMergeKey {
+		mergedParamSets[i] = mergedParamSet
+		i += 1
+	}
+
+	return mergedParamSets, nil
+}
+
+func getParamSetsByMergeKey(mergeKeys []string, paramSets []map[string]string) (map[string]map[string]string, error) {
+	if len(mergeKeys) < 1 {
+		return nil, NoMergeKeys
+	}
+
+	deDuplicatedMergeKeys := make(map[string]bool, len(mergeKeys))
+	for _, mergeKey := range mergeKeys {
+		deDuplicatedMergeKeys[mergeKey] = false
+	}
+
+	paramSetsByMergeKey := make(map[string]map[string]string, len(paramSets))
+	for _, paramSet := range paramSets {
+		paramSetKey := make(map[string]string)
+		for mergeKey, _ := range deDuplicatedMergeKeys {
```

Comment:
Go allows you omit the second value if you are just interested in the first one:
This should work the same way:
```Go
for mergeKey := range deDuplicatedMergeKeys {
```